Expand properties in SdkReference on evaluation

andrew-boyarshin opened the pull request:
Fix #5349.

Unlike #5378 (which didn't fix #5349 but introduced a lot of refactorings around `SdkReference`), this doesn't do anything but implement the fix for #5349. #5378 is absolutely separate from this PR and the feature request #5349 in question.

I emphasize the simplicity of adding this feature, and the flexibility it provides.

andrew-boyarshin commented:
Uhm... My added test (`Microsoft.Build.UnitTests.OM.Construction.ProjectSdkImplicitImport_Tests.SdkPropertiesAreExpandedBeforeResolution`) is failing on CI, but passing locally.
![image](https://user-images.githubusercontent.com/3063662/88626949-b9151700-d0d5-11ea-9663-19e00e2f9b1e.png)
![image](https://user-images.githubusercontent.com/3063662/88627004-d21dc800-d0d5-11ea-83aa-ec0aa8c16a08.png)

Where do the extra messages come from? How can I debug this? Trying to add more logging now...

cdmihai commented on src/Build/Evaluation/Evaluator.cs:
```diff
@@ -1790,8 +1791,53 @@ private List<ProjectRootElement> ExpandAndLoadImports(string directoryOfImportin
+
+                    if (!Equals(expandedSdkReference, sdkReference))
+                    {
+                        var eventArgs = new ProjectImportedEventArgs(
```
AFAIK msbuild does not log expansions, in which case it's probably best to be consistent and not log this.s

andrew-boyarshin replied:
But MSBuild logs e.g. property overrides, which really helps debug cases when something expected is overridden with sth unexpected. I can remove this, of course (that would also drop the new string in resx), but I tried to make this a bit more debuggable.

cdmihai replied:
I agree with the debuggability part, however the following arguments make me not want the log event for this case:
- other expansions are not logged. If we decide to do something else, which I agree might improve debuggability, I'd rather we have a unified approach that logs all expansions. That's a bigger project for a different PR, and needs careful design to avoid perf issues or log size inflation.
- the expanded sdk reference information already kinda gets indirectly logged when the import on the expanded sdk reference gets done. There is room for improvement here: the ProjectImportedEventArgs does not contain information on whether the import was an sdk import and if it was, what the SdkReference contents was. It's a somewhat unrelated change, if you do it you can open a new PR with it, rebased on this current PR, and we'll merge them in cadence. But I think it would be super useful for debugging to be able to search the binlog for all the sdk based imports.

andrew-boyarshin commented:
I've rebased my branch on latest master and applied suggestion from @cdmihai (removed logging). I've also fixed confirmation biased tests by introducing tests that actually test failure cases.

cdmihai commented on src/Build/Evaluation/Evaluator.cs:
```diff
@@ -1790,8 +1791,33 @@ private List<ProjectRootElement> ExpandAndLoadImports(string directoryOfImportin
+
+                    var sdkReferenceOrigin = importElement.SdkLocation;
+
+                    sdkReference = new SdkReference(
```
Before you had a check to only create a `new SdkReference` if expansion changed the contents of the original sdk reference string. I think that's a sensible check to keep.

andrew-boyarshin replied:
The check below that I've removed only influenced logging the event. It didn't influence overriding `sdkReference` variable. When `sdkReference` contains something that can theoretically be expanded (`$(` heuristic above), it is always created anew from the result of the expansion.

cdmihai replied:
Ah, yes, good point.

Forgind commented on src/Build/Evaluation/Evaluator.cs:
```diff
@@ -1779,7 +1779,8 @@ private List<ProjectRootElement> ExpandAndLoadImports(string directoryOfImportin
             string project = importElement.Project;
 
-            if (importElement.SdkReference != null)
+            var sdkReference = importElement.SdkReference;
```
nit:
Real type

Forgind commented on src/Build/Evaluation/Evaluator.cs:
```diff
@@ -1790,8 +1791,33 @@ private List<ProjectRootElement> ExpandAndLoadImports(string directoryOfImportin
                 if (solutionPath == "*Undefined*") solutionPath = null;
                 var projectPath = _data.GetProperty(ReservedPropertyNames.projectFullPath)?.EvaluatedValue;
 
+                var compareInfo = CultureInfo.InvariantCulture.CompareInfo;
```
nit:
real type

Forgind commented on src/Build/Evaluation/Evaluator.cs:
```diff
@@ -1790,8 +1791,33 @@ private List<ProjectRootElement> ExpandAndLoadImports(string directoryOfImportin
+                {
+                    static string EvaluateProperty(string value, IElementLocation location, Expander<P, I> expander) =>
+                        value != null
+                            ? expander.ExpandIntoStringAndUnescape(value,
```
I could somewhat imagine this being problematic if you want your name to include a %, and it wasn't escaping it previously. If your name includes $(, you're probably asking for it if you don't expect it to be expanded. Is it feasible for it to be a path intended for use as a key at some point? Because ExpandIntoStringAndUnescape also fixes file paths.

andrew-boyarshin replied:
You propose using `ExpandIntoStringLeaveEscaped` instead, right? So that it would preserve special characters, like `%`, not treat them like a part of escaped character sequence. I'm not sure about this part, I leave it for MSBuild team to decide. If you think it's better to use the other option, I'll gladly make the change.

cdmihai replied:
@Forgind, trying to summarize the failure cases. Let me know if I've missed any:
- user has a `$(` which they don't want expanded, but now it will get expanded. The traditional approach here is for the user to escape any msbuild syntax when they want it as a literal, which will still work here.
- on non windows systems, the `$(` is [used as a guard to prevent adjusting path like strings](https://github.com/cdmihai/msbuild/blob/9c2d922959e80ba72fefb7addec21ede227639c3/src/Shared/FileUtilities.cs#L454). So unix based users with strings that contained both `$(` and backslashes in their sdk references, will now get their slashes changed. I think this is still fine:
  - first, `$(` and slashes would constitute illegal nuget names and versions, so their builds would have probably crashed anyway had they tried to put these in nuget based sdks.
  - second, even though backslashes are valid unix filename characters, AFAIK their usage is not common. And even if the user did have backslashes in their name / version, they can escape them. If this gets to be a common case (and thus make escaping bad for usability) we could then add an option to not fix the file path when expanding the sdk reference (my guess is it won't happen).

Forgind replied:
These are certainly all things the user can get out of pretty easily; I'm just trying to avoid having code that works with one MSBuild version, then fails with the next. I was also thinking of the failure case @andrew-boyarshin mentioned with %. The most odd part to me is if you have, say, a name with an escape character but no properties. Then you add a property to the version, and the name is escaped. That also applies to paths, so you noted that slashes would be illegal—are quotes? Since MaybeAdjustFilePath also removes quotes if present. I think that was all I was thinking. I'm not too worried about most properties being expanded, but it did occur to me that if someone uses the guard against paths being adjusted, it might look like a property and be converted to "" when it can't be found. That's the worst case I can imagine.

cdmihai replied:
> trying to avoid having code that works with one MSBuild version, then fails with the next

If it does break anybody, it's probably in their own interest to fix it, since the correct behaviour from the very start should have been to expand the sdk reference. So they'd be migrating off of relying on a bug.

However, in case it does break anybody, it probably makes sense to enable the expansion by default but provide an escape hatch so if anybody does break, they can revert to the old behaviour. @andrew-boyarshin, can you please add an escape hatch [here](https://github.com/dotnet/msbuild/blob/f365aa332f58cf43e78d76491438f15fc1988f27/src/Shared/Traits.cs#L110), and disable sdk reference expansion if the escape hatch is enabled?

>MaybeAdjustFilePath also removes quotes if present

I don't think it does. The results from RemoveQuotes does not seem to flow into the return of MaybeAdjustFilePath

>a name with an escape character but no properties. Then you add a property to the version, and the name is escaped.

This breaking scenario [is constrained ](https://github.com/cdmihai/msbuild/blob/9c2d922959e80ba72fefb7addec21ede227639c3/src/Deprecated/Engine/Shared/EscapingUtilities.cs#L65-L66) only to strings that have `%` followed by two digits. In which case (beyond the fact nuget would probably fail on it), the user can use the escape hatch.

andrew-boyarshin commented:
```xml
<Project Sdk="MSBuild.Sdk.Extras/$(MySdkVersion)">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net5.0</TargetFramework>
    <RootNamespace>msbuild_test</RootNamespace>
  </PropertyGroup>

</Project>
```

![image](https://user-images.githubusercontent.com/3063662/89095238-2686b780-d3f6-11ea-821a-030f374f8482.png)
![image](https://user-images.githubusercontent.com/3063662/89095243-38685a80-d3f6-11ea-9bc1-b3cd10677ae0.png)
![image](https://user-images.githubusercontent.com/3063662/89095255-57ff8300-d3f6-11ea-91ed-10f56d888381.png)
![image](https://user-images.githubusercontent.com/3063662/89095258-6352ae80-d3f6-11ea-8cb9-74ca3d79c0bc.png)
![image](https://user-images.githubusercontent.com/3063662/89095263-782f4200-d3f6-11ea-9847-065ca48c0d46.png)
![image](https://user-images.githubusercontent.com/3063662/89095267-82e9d700-d3f6-11ea-85e7-5f18427f8a35.png)
![image](https://user-images.githubusercontent.com/3063662/89095278-8b421200-d3f6-11ea-8c25-fa3ac063506c.png)

Forgind commented on src/Build/Evaluation/Evaluator.cs:
```diff
@@ -1790,8 +1792,58 @@ private List<ProjectRootElement> ExpandAndLoadImports(string directoryOfImportin
+                                    return expander.ExpandIntoStringAndUnescape(value, Options, location);
+                                case SdkReferencePropertyExpansionMode.ExpandLeaveEscaped:
+                                    return expander.ExpandIntoStringLeaveEscaped(value, Options, location);
+                                case SdkReferencePropertyExpansionMode.NoExpansion:
```
Can you add test cases for the new modes? In particular, without a break in here, I believe it will fall through on the `NoExpansion` and `DefaultExpand` cases to throw an error.

andrew-boyarshin replied:
But that's intentional. Both cases are unreachable here, and must cause an internal assertion error. I've listed all enum values to make it explicit and make it more future-proof in case new enum values are added.

andrew-boyarshin replied:
Additionally, I'm not sure I can test it. EscapeHatches do the read only once in a process lifetime, so one must make sure that each test restarts the test host. Is there any existing example in MSBuild tests on how to do it?

andrew-boyarshin replied:
Okay, it seems xUnit unloads and reloads the assemblies for each subtest, so the EscapeHatches takes the correct value on each subtest..

Forgind commented on src/Build/Evaluation/Evaluator.cs:
```diff
@@ -1790,8 +1792,58 @@ private List<ProjectRootElement> ExpandAndLoadImports(string directoryOfImportin
+                                case SdkReferencePropertyExpansionMode.NoExpansion:
+                                case SdkReferencePropertyExpansionMode.DefaultExpand:
+                                default:
+                                    ErrorUtilities.ThrowArgumentOutOfRange(nameof(mode));
```
This should never happen, right? Maybe throw an internal error rather than ArgumentOutOfRange.

andrew-boyarshin replied:
Initially I had a `switch` expression that was throwing `ArgumentOutOfRangeException`. After I've discovered there is a matching `ErrorUtilities` method, I've rewritten it to be a `switch` statement and use that. Rationale: however unlikely case this is, it's better to have an opportunity to prevent throwing and use a fallback.

Forgind replied:
Oh, missed that. You're right.

andrew-boyarshin commented:
I've greatly rewritten tests so that both success and failure cases are tested within a single parametrized test method. I've also increased the number of tested value combinations: from 30 manually specified cases up to 240 automatically-generated ones.

Forgind commented on src/Build.UnitTests/Evaluation/ProjectSdkImplicitImport_Tests.cs:
```diff
@@ -534,6 +542,257 @@ public void IgnoreMissingImportsSdkNotFoundDoesNotThrow()
+            }
+        }
+
+        [Theory]
```
nit:
Does this still need to be a theory?

andrew-boyarshin replied:
Well, since it is a parametrized test, it has to be a `Theory`, not a `Fact`.

Forgind replied:
I was thinking there's only one input used with the test, so you could inline it.

andrew-boyarshin replied:
You propose inlining the whole set of subtests from `SdkPropertiesAreExpandedData` into this method? I'm not sure why is that better than having xUnit do the full stats per-subtest.

andrew-boyarshin replied:
I'm also not sure what you mean by "one input". That's 240 different dynamically-generated inputs.

Forgind replied:
I should have said the data come from one place. That means that converting it to a Fact would mean a single for loop, which isn't too messy. I normally think theories are better if I would need to specify a lot of data in the test to keep it a Fact. In this case, you could remove the `new object[]` part in SdkPropertiesAreExpandedData and make it a simple loop through all the Cases, which isn't ugly.

Forgind replied:
In any case, it isn't bothering me too much, so you can ignore me if you want.

andrew-boyarshin replied:
1. Will setting the environment variable still work? The value in `EscapeHatches` is set on first access, it can't be changed later, so each change in Mode requires at least reloading the assemblies (if not restarting the test host altogether).
2. Why reimplement the xUnit parametrized test? What's the expected behavior if one subtest fails? Assert will stop the iteration and the log will contain only the fact that something has failed, not the data on which it failed or the amount of successful/failed subtests.

I strongly oppose the suggested inlining.

Forgind replied:
I actually don't know about (1), so if it doesn't, you're probably right. I would have thought `var project = Project.FromXmlReader(xmlReader, projectOptions);` is the point at which the environment variable needs to have changed, since that simulates making a new project.
For (2), I tend to debug into them the same, so it doesn't really matter to me.

You feel far more strongly than I do, so it's fine to leave it as-is. I already approved.

cdmihai commented on src/Build.UnitTests/Evaluation/ProjectSdkImplicitImport_Tests.cs:
```diff
@@ -534,6 +542,257 @@ public void IgnoreMissingImportsSdkNotFoundDoesNotThrow()
+                    yield return SdkReferencePropertyExpansionMode.ExpandLeaveEscaped;
+                }
+
+                static IEnumerable<(SdkPropertiesAreExpandedDataTemplate, bool, bool)> Templates()
```
nit: name the tuple elements to self document the two bools

andrew-boyarshin replied:
I had them initially documented, but then the hard wrap at 120 characters made the function definition look bad, and I decided since this method is used only once (and can only be used within this property), and the return value is immediately deconstructed with meaningful variable names, this can be left with default tuple field names. I can add the tuple field names, but this'll make method definition look worse. Your call.

cdmihai replied:
As middle ground you can only name the two bools (`setName, setVersion`, not that long) and leave the first tuple element unnamed, since the type is expressive already.

BenVillalobos commented:
Seeing as https://github.com/dotnet/msbuild/issues/5646 hit many other PRs I'd like to close and reopen this just in case.
